File: 2024/01/part1.cpp
#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <vector>

constexpr const char* INPUT_FILE = "input.txt";
using value_type = int;

ulong distance(std::vector<value_type>& first, std::vector<value_type>& second) {
    assert(first.size() == second.size());
    ulong sum = 0;

    for (size_t i = 0; i < first.size(); ++i) {
        sum += abs(first[i] - second[i]);
    }

    return sum;
}

void load_input(std::vector<value_type>& first, std::vector<value_type>& second) {
    std::ifstream file(INPUT_FILE); // Open file for reading
    value_type num1, num2;

    while (file >> num1 >> num2) {
        first.push_back(num1);
        second.push_back(num2);
    }
    file.close();
}

int main() {
    std::vector<value_type> first { };
    std::vector<value_type> second { };

    // std::cerr << "loading input" << std::endl;
    load_input(first, second);
    // std::cerr << "input loaded" << std::endl;

    std::sort(first.begin(), first.end());
    std::sort(second.begin(), second.end());


    std::cout << distance(first, second) << std::endl;
}

File: 2024/04/part1.py
def get_at_inx(input, x, y):
    if (x < 0 or len(input[0]) <= x):
        return '.'

    if (y < 0 or len(input) <= y):
        return '.'

    return input[y][x]

def test_possition(input, x, y):
    if get_at_inx(input, x, y) != 'A':
        return False

    s1 = { get_at_inx(input, x - 1, y - 1), get_at_inx(input, x + 1, y + 1) }
    s2 = { get_at_inx(input, x + 1, y - 1), get_at_inx(input, x - 1, y + 1) }

    return 'M' in s1 and 'M' in s2 and 'S' in s1 and 'S' in s2


def search_with_vector(input):
    sum = 0

    for x in range(len(input[0])):
        for y in range(len(input)):
            if test_possition(input, x, y):
                sum += 1


    return sum


with open("input.txt", 'r') as file:
    input = file.readlines()

sum = 0

sum += search_with_vector(input);

print(sum)

